Subscribing to notifications while inside the said instrumented section

thedarkone opened the pull request:
@skaestle posted a repro case for #21873, though the post has now been deleted:

> @sgrif hello! would you mind reopening this issue? I have created a small rails (4.2.5) app that showcases the error: https://github.com/skaestle/investigate-rails-notification.
> 
> this happens when an error is thrown while subscribed to `process_action.action_controller`: https://github.com/skaestle/investigate-rails-notification/blob/master/app/controllers/posts_controller.rb#L7

The issue is that subscribing to a notification while inside the said instrumented section sometimes results in an error.

```
ActiveSupport::Notifications.instrument('foo') do
  ActiveSupport::Notifications.subscribe('foo') {}
end
```

I'm not sure if such a use case should be supported, but I'm submitting a PR anyways. I **cannot** rule out that a legitimate use case exists: ie for example, if the loading of Rails Engines was instrumented and an Engine (while being loaded) subscribed to the notification, it might error out.

sgrif commented:
I think it's fine to go ahead and support this, but I'd rather just have all of those comments in the commit message instead.

thedarkone committed:
Subscribing to notifications while inside the said instrumented section.

The issue is that on the exit from Instrumenter#instrument section,
an Evented listener will run into an error because its thread local
(Thread.current[:_timestack]) has not been set up by the #start
method (this obviously happens because the Evented listeners didn't
exist at the time, since no subscribtion to that section was made yet).

Note: support for subscribing to instrumented sections, while being
inside those instrumented sections, might be removed in the future.

Maybe fixes #21873.

thedarkone commented:
Re-pushed.

eileencodes commented:
Thanks @thedarkone!
